src/cockpit/polkit.py

```diff
@@ -0,0 +1,140 @@
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if line.startswith(b'PAM_PROMPT'):
+                prompt = line.split(b' ', 1)[1].strip().decode()
+                # Don't pass this to the UI if it's "Password:" (the usual case),
+                # so that superuser.py uses the translated default
+                if prompt.startswith('Password'):
+                    prompt = ''
+
+                try:
+                    passwd = await self.responder.do_askpass('', prompt, '')
+                    logger.debug('got PAM_PROMPT %s, do_askpass returned a password', prompt)
+                except asyncio.CancelledError:
+                    logger.debug('got PAM_PROMPT %s, do_askpass cancelled', prompt)
+                    process.terminate()
+                    break
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif line.startswith(b'PAM_TEXT_INFO'):
+                logger.info(line.decode())
+            elif line.startswith(b'PAM_ERROR'):
+                logger.warning(line.decode())
+            elif line.startswith(b'SUCCESS'):
+                logger.info('Authentication succeeded')
+                break
+            elif line.startswith(b'FAILURE'):
+                logger.warning('Authentication failed')
+                break
+            else:
+                logger.warning('Unknown line from helper, aborting: %s', line)
+                process.terminate()
+                break
+
+        res = await process.wait()
+        logger.debug('helper exited with code %i', res)
+
+    @bus.Interface.Method('', 's')
+    def cancel_authentication(self, cookie):
+        logger.warning('CancelAuthentication: %s not implemented', cookie)
+
+
+def register_agent(responder):
+    '''Register polkit agent.
+
+    Returns an opaque object which needs to be kept during the lifetime of the agent.
+    After that, pass it to @unregister_agent().
+    '''
+    try:
+        system_bus = bus.Bus.default_system()
+    except FileNotFoundError:
+        logger.debug('system bus not found, not registering polkit agent')
+        return None
+
+    try:
+        # may refine that with a D-Bus call to logind
+        subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
+    except KeyError:
+        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
+        return None
+
+    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
+    agent_slot = system_bus.add_object('/', agent_object)
+    logger.debug('added agent object %s', agent_slot)
+
+    # register agent
+    (loc, encoding) = locale.getlocale()
+    locale_name = loc or 'C'
+    if encoding:
+        locale_name += '.' + encoding
```

Comment:
This is terrible. See https://docs.python.org/3/library/locale.html#locale.getlocale

Perhaps there is a better API to get something like `de_DE.UTF-8`?

Also, we mostly don't care -- we handle translation for the standard "Password:" case ourselves, and this stuff would still fail pretty badly with something like 2FA anyway.

Reply:
The [glibc documentation](https://www.gnu.org/software/libc/manual/html_node/Locale-Names.html) at least confirms that the `.` is fixed for separating the encoding. However, this is bad for the non-UTF8 cases:
```
❱❱❱ LANG=de_DE python3 -c 'import locale; print(locale.getlocale())'
('de_DE', 'ISO8859-1')

❱❱❱ LANG=de_DE@euro python3 -c 'import locale; print(locale.getlocale())'
('de_DE', 'ISO8859-15')
```

`LANG=de_DE.iso8859-15@euro locale` actually works, but (1) that is not the canonical locale name, and (2) that variant isn't handled  here.

Reply:
```python
>>> libc = ctypes.CDLL(None)
>>> libc.setlocale.restype = ctypes.c_char_p
>>> libc.setlocale(locale.LC_MESSAGES, '')
b'en_CA.UTF-8'
```

eh...

Reply:
But also:

```python
>>> locale.setlocale(locale.LC_MESSAGES, '')
'en_CA.UTF-8'
```